veps/sig-compute/10-dra-devices/vep.md

```diff
@@ -0,0 +1,785 @@
+kind: VirtualMachineInstance
+metadata:
+  name: vmi-fedora
+  namespace: gpu-test1
+spec:
+  resourceClaims:
+  - name: gpu-resource-claim
+    resourceClaimTemplateName: single-gpu
+  domain:
+    gpus:
+    - claimName: gpu-resource-claim
+      requestName: gpu
+      name: example-pgpu
+status:
+  deviceStatus:
+    gpuStatuses:
+    - name: example-pgpu
+      deviceResourceClaimStatus:
+        name: gpu-0
+        resourceClaimName: virt-launcher-vmi-fedora-hhzgn-gpu-resource-claim-c26kh
+        attributes:
+          pciAddress: 0000:01:00.0
+```
+
+## Alternatives
+
+## Alternative 1
+```
+type GPU struct {
+	// Name of the GPU device as exposed by a device plugin
+	Name              string       `json:"name"`
+	DeviceName        string       `json:"deviceName"`
+	VirtualGPUOptions *VGPUOptions `json:"virtualGPUOptions,omitempty"`
+	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
+	// +optional
+	Tag   string `json:"tag,omitempty"`
+	Claim string `json:"claim,omitempty"`
+}
+
+type HostDevice struct {
+	Name string `json:"name"`
+	// DeviceName is the resource name of the host device exposed by a device plugin
+	DeviceName string `json:"deviceName"`
+	// If specified, the virtual network interface address and its tag will be provided to the guest via config drive
+	// +optional
+	Tag string `json:"tag,omitempty"`
+	// If specified, the ResourceName of the host device will be provisioned using DRA driver . which will not require the deviceName field
+	//+optional
+	ResourceClaim *k8sv1.ResourceClaim `json:"resourceClaim,omitempty"`
+}
+
+// ResourceClaim represents a resource claim to be used by the virtual machine
+type ResourceClaim struct {
+	// Name is the name of the resource claim
+	Name string `json:"name"`
+	// Source represents the source of the resource claim
+	Source ResourceClaimSource `json:"source"`
+}
+
+// ResourceClaimSource represents the source of a resource claim
+type ResourceClaimSource struct {
+	// ResourceClaimName is the name of the resource claim
+	ResourceClaimName string `json:"resourceClaimName"`
+	// ResourceClaimTemplateName is the name of the resource claim template
+	//
+	// Exactly one of ResourceClaimName and ResourceClaimTemplateName must
+	// be set.
+	ResourceClaimTemplateName string `json:"resourceClaimTemplateName"`
+}
+```
+This design misses the use-case where more than one DRA device is specified in the claim template, as each
+device will have its own template in the API.
+This design also assumes that the deviceName will be provided in the ClaimParameters, which requires the DRA drivers
+to have a ClaimParameters.spec.deviceName in their spec.
+
+## Alternative 2
+Asking the dra plugin authors to inject env variable to CDI spec.
+In order to uniquely identify the device required by the vmi spec, the follow env variable will have to be constructed:
+```
+PCI_RESOURCE_<RESOURCE-CLAIM-NAME>_<REQUEST-NAME>="0000:01:00.0"
```

Comment:
I don't think the RESOURCE-CLAIM-NAME is needed here. This can continue to be PCI_RESOURCE_<RESOURCE_NAME> or PCI_RESOURCE_<requestName> I don't see why a reference to claim is needed.

Reply:
1. `PCI_RESOURCE_<RESOURCE_NAME>`

I assume by RESOURCE_NAME you mean the GPU name in the following spec:
```
spec:
  resourceClaims:
  ...
  domain:
    devices:
      gpus:
      - name: pgpu
```

so the proposed env var will be PCI_RESOURCE_PGPU here.

This will not work because vmi spec is not available in dra plugin, DRA plugin authors will not have any idea of what name VMI user has set for the device

2. `PCI_RESOURCE_ ` this will not work because there are multiple devices in the same resource claim, and each device could be from a different driver, how can virt-launcher know which device is PCI resource and which one is MDEV, even this will not work